test/drbgtest.c

```diff
@@ -864,13 +865,79 @@ static int test_multi_thread(void)
 }
 #endif
 
+/*
+ * This function is only returns the entropy already added with RAND_add(),
+ * and does not get entropy from the OS.
+ */
+static size_t rand_add_entropy(RAND_DRBG *drbg,
+                               unsigned char **pout,
+                               int entropy, size_t min_len, size_t max_len,
+                               int prediction_resistance)
+{
+    size_t ret = 0;
+
+    if (drbg->pool == NULL)
+        return 0;
+
+    if (drbg->pool->entropy < (size_t)entropy || drbg->pool->len < min_len
+        || drbg->pool->len > max_len)
+        return 0;
+
+    *pout = malloc(drbg->pool->len);
+    if (*pout == NULL)
+        return 0;
+
+    memcpy(*pout, drbg->pool->buffer, drbg->pool->len);
+    ret = drbg->pool->len;
+    OPENSSL_secure_clear_free(drbg->pool->buffer, drbg->pool->max_len);
+    OPENSSL_free(drbg->pool);
+    drbg->pool = NULL;
+
+    return ret;
+}
+
+static void rand_add_clean_entropy(RAND_DRBG *drbg, unsigned char *out, size_t outlen)
+{
+    free(out);
+}
+
+/*
+ * Test that instantiating works when OS entropy is not available and that
+ * RAND_add() is enough to reseed it.
+ */
+static int test_rand_add()
+{
+    RAND_DRBG *master = RAND_DRBG_get0_master();
+    RAND_DRBG_get_entropy_fn old_get_entropy = master->get_entropy;
+    RAND_DRBG_cleanup_entropy_fn old_cleanup_entropy = master->cleanup_entropy;
+    int rv = 0;
+    unsigned char rand_add_buf[256];
+
+    master->get_entropy = rand_add_entropy;
+    master->cleanup_entropy = rand_add_clean_entropy;
+    master->reseed_counter++;
+    RAND_DRBG_uninstantiate(master);
+    memset(rand_add_buf, 0, sizeof(rand_add_buf));
```

Comment:
Better memset the buffer to a nonzero value like0xCD (see my patch). It does not make a difference for the test itself, but for the developer how tries to verify the test implementation in the debugger.